bin/proving-service/src/proxy/status.rs

```diff
@@ -0,0 +1,128 @@
+use std::{sync::Arc, time::Duration};
+
+use pingora::{
+    apps::HttpServerApp,
+    http::ResponseHeader,
+    prelude::*,
+    protocols::{Stream, http::ServerSession},
+    server::ShutdownWatch,
+};
+use serde::Serialize;
+use tonic::{async_trait, transport::Channel};
+use tracing::error;
+
+use super::worker::WorkerHealthStatus;
+use crate::{
+    commands::worker::ProverType, error::ProvingServiceError,
+    generated::status::status_api_client::StatusApiClient, proxy::LoadBalancerState,
+};
+
+/// Create a gRPC [StatusApiClient] for the given worker address.
+///
+/// # Errors
+/// - [ProvingServiceError::InvalidURI] if the worker address is invalid.
+/// - [ProvingServiceError::ConnectionFailed] if the connection to the worker fails.
+pub async fn create_status_client(
+    address: String,
+    connection_timeout: Duration,
+    total_timeout: Duration,
+) -> Result<StatusApiClient<Channel>, ProvingServiceError> {
+    let channel = Channel::from_shared(format!("http://{}", address))
+        .map_err(|err| ProvingServiceError::InvalidURI(err, address.clone()))?
+        .connect_timeout(connection_timeout)
+        .timeout(total_timeout)
+        .connect()
+        .await
+        .map_err(|err| ProvingServiceError::ConnectionFailed(err, address))?;
+
+    Ok(StatusApiClient::new(channel))
+}
+
+// Status of a worker
+#[derive(Debug, Serialize)]
+pub struct WorkerStatus {
+    address: String,
+    version: String,
+    status: WorkerHealthStatus,
+}
+
+/// Status of the proxy
+#[derive(Debug, Serialize)]
+pub struct ProxyStatus {
+    version: String,
+    supported_proof_type: ProverType,
+    busy_workers: usize,
+    workers: Vec<WorkerStatus>,
+}
+
+/// Service that handles status requests
+pub struct StatusService {
+    load_balancer: Arc<LoadBalancerState>,
+}
```

Comment:
This is specifically for the proxy, right? if so, I'd call it `ProxyStatusService`.

Reply:
Yes, it is. I renamed it.